drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -282,6 +316,59 @@ static int siwx91x_status(const struct device *dev, struct wifi_iface_status *st
 	return 0;
 }
 
+static int siwx91x_check_current_iface(sl_wifi_interface_t iface, uint8_t z_mode)
+{
+	switch (z_mode) {
+	case WIFI_STA_MODE:
+		if (FIELD_GET(SIWX91X_INTERFACE_MASK, iface) == SL_WIFI_CLIENT_INTERFACE) {
+			return -EALREADY;
+		}
+		break;
+	case WIFI_AP_MODE:
+		if (FIELD_GET(SIWX91X_INTERFACE_MASK, iface) == SL_WIFI_AP_INTERFACE) {
+			return -EALREADY;
+		}
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int siwx91x_mode(const struct device *dev, struct wifi_mode_info *mode)
+{
+	sl_wifi_device_configuration_t network_config;
+	struct siwx91x_dev *sidev = dev->data;
+	sl_wifi_interface_t interface;
+	int ret;
+
+	__ASSERT(mode, "mode cannot be NULL");
+
+	siwg917_nwp_get_configuration(&network_config);
+	interface = sl_wifi_get_default_interface();
+
+	ret = siwx91x_check_current_iface(interface, mode->mode);
+	if (ret) {
+		return ret;
+	}
+
+	if (siwx91x_get_mode(mode->mode) < 0) {
+		return -EINVAL;
```

Comment:
If `siwx91x_get_mode() < 0`, it means that we don't have to reboot, so we should return 0, no?

BTW, I believe the following code would be more explicit for the reader:
```
    int current_mode = siwx91x_get_iface_mode(interface);
    int requested_mode = siwx91x_convert_mode(mode->mode);
    if (current_mode == requested_mode) {
        return 0;
    }
    siwx91x_nwp_reboot(requested_mode);
```